pkg/controller/pod.go

```diff
@@ -0,0 +1,246 @@
+}
+
+func (r *ReconcilePod) getClient() client.Client {
+	return r.client
+}
+func (r *ReconcilePod) getAzClient() azClientSet.Interface {
+	return r.azVolumeClient
+}
+func (r *ReconcilePod) getNamespace() string {
+	return r.namespace
+}
+
+// Implement reconcile.Reconciler so the controller can reconcile objects
+var _ reconcile.Reconciler = &ReconcilePod{}
+
+func (r *ReconcilePod) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	var pod corev1.Pod
+	var value interface{}
+	var exists bool
+
+	podKey := getNamespacedName(request.Namespace, request.Name)
+	if err := r.client.Get(ctx, request.NamespacedName, &pod); err != nil {
+		// if the pod has been deleted, remove entry from podToClaimsMap and claimToPodMap
+		if errors.IsNotFound(err) {
+			value, exists = r.controllerSharedState.podToClaimsMap.Load(podKey)
+			if !exists {
+				return reconcile.Result{}, nil
+			}
+			claims := value.([]string)
+
+			for _, claim := range claims {
+				r.controllerSharedState.claimToPodsMap.Delete(claim)
+			}
+			r.controllerSharedState.podToClaimsMap.Delete(podKey)
+			return reconcile.Result{}, nil
+		}
+		klog.Errorf("failed to get Pod (%s): %v", podKey, err)
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	// update the shared map
+	r.updateMap(&pod)
+
+	// create replica batches
+	if err := r.createReplicas(ctx, &pod); err != nil {
+		return reconcile.Result{Requeue: true}, err
+	}
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcilePod) createReplicas(ctx context.Context, pod *corev1.Pod) error {
+	if pod == nil {
+		return nil
+	}
+	podKey := getNamespacedName(pod.Namespace, pod.Name)
+	volumes, err := r.controllerSharedState.getVolumesFromPod(podKey)
+	if err != nil {
+		return err
+	}
+
+	nodes, err := getNodesForReplica(ctx, r, r.controllerSharedState, volumes, podKey)
+	if err != nil {
+		return err
+	}
+
+	for _, volume := range volumes {
+		azVolume, err := azureutils.GetAzVolume(ctx, r.client, r.azVolumeClient, volume, r.namespace, true)
+		if err != nil {
+			return err
+		}
+		// if underlying volume is not present, abort operation
+		if !isCreated(azVolume) {
+			return status.Errorf(codes.Aborted, "azVolume (%s) has no underlying volume object", azVolume.Name)
+		}
+
+		// get all replica attachments for the given volume
+		replicaNodes, err := getNodesWithReplica(ctx, r, volume)
+		// if there already are replica attachments for the volume, let the replica reconciler handle replica creation and skip batch creation to avoid race between two controllers
+		if len(replicaNodes) > 0 {
+			continue
+		} else if err != nil {
+			return err
+		}
+
+		numCreated := 0
+		for _, node := range nodes {
+			if numCreated >= azVolume.Spec.MaxMountReplicaCount {
+				break
+			}
+			if err := createReplica(ctx, r, azVolume.Status.Detail.ResponseObject.VolumeID, node); err != nil {
+				return err
+			}
+			numCreated++
+		}
+	}
+	return nil
+}
+
+func (r *ReconcilePod) updateMap(pod *corev1.Pod) {
```

Comment:
Consider making this a method of `SharedState` since that is what it operates on.